classes/firmware-qcom-sign.bbclass

```diff
@@ -0,0 +1,133 @@
+		vpu30_4v_16mb.mbn VENUS-FW 0 ${newline}\
+		wpss.mbn WPSS 1 ${newline}\
+		xbl_config.elf XBL-CONFIG 0 ${newline}\
+		xbl_config_gunyah.elf XBL-CONFIG 0 ${newline}\
+		xbl_config_kvm.elf XBL-CONFIG 0 ${newline}\
+		xbl.elf XBL 0 ${newline}\
+		XblRamdump.elf XBL-RAM-DUMP 0 ${newline}\
+		DigestsToSign.bin.mbn VIP 0 ${newline}\
+		FD02C9DA-306C-48C7-A49C-BBD827AE86EE.mbn TZ-APP-OEM 0 ${newline}\
+	"
+
+	sign_id=$(${sectools} secure-image --inspect $1 | grep "| Software ID:" | cut -d'|' -f3)
+	if [ -z "${sign_id}" ]; then
+		bbwarn "$file does not contain signatures, skipping"
+		return 0
+	fi
+
+	# Lookup the IMAGE-ID from mapping
+	image_id=$(echo "${image_id_mapping}" | grep "${file}" | cut -d' ' -f2)
+	bbdebug 1 "IMAGE_ID = ${image_id}"
+	if [ -z "${image_id}" ] || [ "${image_id}" = "UNKNOWN" ]; then
+		bbfatal "Unable to find IMAGE-ID mapping for ${file}"
+	fi
+
+	if [ "${image_id}" = "SKIP" ]; then
+		bbwarn "Skip signing of ${file}"
+		return 0
+	fi
+
+	bbnote "Signing ${1}: ${sectools} secure-image \
+		--sign $1 --image-id=${image_id} \
+		--security-profile ${secprofile} \
+		--anti-rollback-version=0x0 \
+		--signing-mode LOCAL \
+		--root-certificate-index 0 \
+		--root-certificate=${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_rootca0.cer \
+		--ca-certificate=${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_attestca0.cer \
+		--ca-key=${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_attestca0.key \
+		--outfile $1"
+
+	${sectools} secure-image \
+		--sign $1 --image-id=${image_id} \
+		--security-profile ${secprofile} \
+		--anti-rollback-version=0x0 \
+		--signing-mode LOCAL \
+		--root-certificate-index 0 \
+		--root-certificate=${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_rootca0.cer \
+		--ca-certificate=${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_attestca0.cer \
+		--ca-key=${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_attestca0.key \
+		--outfile $1
+	if [ $? -ne 0 ]; then
+		bbfatal "Signing ${file} failed"
+	fi
+
+	bbnote "Verifying root hash of ${file}"
+	${sectools} secure-image --verify-root ${root_hash} $1
+	if [ $? -ne 0 ]; then
+		bbfatal "Root hash of ${file} failed verification"
+	fi
+}
+
+QCOM_FIRMWARE_SIGN_ENABLE ??= "0"
+
+do_firmware_sign() {
+	if [ "${QCOM_FIRMWARE_SIGN_ENABLE}" = "1" ]; then
+		if [ ! -f "${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_rootca0.cer" ]; then
+			bbfatal "QCOM_FIRMWARE_SIGN_KEY_DIR or Root certificate is invalid"
+		fi
+		if [ ! -f "${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_attestca0.cer" ]; then
+			bbfatal "QCOM_FIRMWARE_SIGN_KEY_DIR or CA certificate is invalid"
+		fi
+		if [ ! -f "${QCOM_FIRMWARE_SIGN_KEY_DIR}/qpsa_attestca0.key" ]; then
+			bbfatal "QCOM_FIRMWARE_SIGN_KEY_DIR or CA key is invalid"
+		fi
+		if [ ! -f "${QCOM_FIRMWARE_SIGN_SECTOOLS_DIR}/sectools" ]; then
+			bbfatal "QCOM_FIRMWARE_SIGN_SECTOOLS_DIR is invalid"
+		fi
+		if [ ! -f "${QCOM_FIRMWARE_SIGN_SECPROFILE}" ]; then
+			bbfatal "QCOM_FIRMWARE_SIGN_SECPROFILE is invalid"
+		fi
+
+		bbnote "Searching for MBN files to sign"
+		file_list=$(find "${D}/" -maxdepth 1 -iname "*.mbn")
```

Comment:
Looks like the bellows for loops could be together in a function which would receive the list of files in its input.